// core.js/db/DB.js
var DB = {
    _db: '',
    _result: '',
    connect: function() {
        if (DB._db === '') {
            var mysql = require('mysql');
            DB._db = mysql.createConnection({
                host: 'localhost',
                user: 'root',
                password: '',
                database: 'phonegap'
            });

            DB._db.connect(function(err) {
                if (err) {
                    console.log(err);
                    return false;
                }
            });
        }
        //console.log('Connected DB');
        return DB._db;
    },
    query: function(sql) {        
        //console.log('Query: ' + DB.connect());
        if (DB.connect()) {
            DB._db.query(sql, function(err, rows) {
                if (err) {
                    console.log('error:' + err);
                }
                DB._result = rows;
                console.log('2. result: ' + DB._result);
            });
        }
    },
    escape: function(param) {        
        if (DB.connect()) {
            return DB._db.escape(param);
        }
    }
};

module.exports = DB;
module.exports.query = function(sql) {
    DB.query(sql);    
};
module.exports.getResult = function(){
    return DB._result;
};
module.exports.escape = function(param) {
    return DB.escape(param);
};
